Zc/workbench transition

dvzacharycutler opened the pull request:
Closes *list issues numbers here*

### Developer Checklist (Definition of Done)

* [ ] Descriptive title for this pull request provided (will be used for release notes later)
* [ ] All acceptance criteria from the issue are met
* [ ] Branch is up-to-date with the branch to be merged with, i.e., develop
* [ ] Code is cleaned up and formatted
* [ ] Code documentation written
* [ ] Unit tests written
* [ ] Tested in Chrome
* [ ] Tested in Firefox
* [ ] Build is successful
* [ ] [Summary of changes](#summary-of-changes) written
* [ ] Wiki documentation written
* [ ] Assign at least one reviewer
* [ ] Assign at least one assignee
* [ ] Add type label (e.g., *bug*, *enhancement*) to this pull request
* [ ] Add next version label (e.g., `next-version: minor`) to this PR following [semver](https://semver.org/)


### Summary of changes

* 

### Screenshots

dvmarkusvogl commented on src/app/workbench/Workbench.tsx:
```diff
@@ -0,0 +1,106 @@
+                  }),
+                );
+
+                // setTimeout(() => {
```
do we need this?

dvzacharycutler replied:
nope, fixed

dvmarkusvogl commented on src/app/workbench/Workbench.tsx:
```diff
@@ -0,0 +1,106 @@
+                );
+              }
+
+              // dispatch(
```
do we need this?

dvzacharycutler replied:
nope, fixed

dvmarkusvogl commented on src/app/workbench/WorkbenchView.tsx:
```diff
@@ -25,12 +25,10 @@ export function WorkbenchView({
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  const views = useMemo(() => () => getVisynView(ordino.workbenches[workbenchIndex].entityId), []);
-  const { value } = useAsync(views, []);
+  const { value } = useAsync(getVisynView, [ordino.workbenches[workbenchIndex].entityId]);
```
bad variable name: value.

dvzacharycutler replied:
ya good point, i actually do this a lot when using that useAsync where I dont bother renaming the returned value. I'll get out of that habit, renamed it to visynViews here.

dvmarkusvogl commented on src/app/workbench/sidebar/CreateNextWorkbenchSidebar.tsx:
```diff
@@ -51,6 +50,8 @@ export function CreateNextWorkbenchSidebar({ workbench }: ICreateNextWorkbenchSi
     return entities;
   }, [status, availableViews]);
 
+  console.log(availableViews);
```
should this be in the final version?

dvzacharycutler replied:
Same as below

dvmarkusvogl commented on src/app/workbench/sidebar/CreateNextWorkbenchSidebar.tsx:
```diff
@@ -126,7 +127,7 @@ export function CreateNextWorkbenchSidebar({ workbench }: ICreateNextWorkbenchSi
                   }}
                 >
-                  {availableViews
+                  {/* {availableViews
```
delete if not needed?

dvzacharycutler replied:
good point, but also made me realize we dont need this whole file anymore, so I deleted it!

dvmarkusvogl commented on src/app/workbench/sidebar/WorkbenchUtilsSidebar.tsx:
```diff
@@ -0,0 +1,69 @@
+        return <div style={{ width: '250px' }}>Comment something</div>;
+      }
+      default: {
+        return <div style={{ width: '250px' }}>Thats weird</div>;
```
More speaking "error case", user has no idea what 'that's weird' implies

dvzacharycutler replied:
Ya definitely true, all of these hard coded components aren't great, and will be changed as we create actual components for them, but that one's particularly bad. Should probably make a "error" type component to normalize this kind of thing in the future, but changed the text to "There was an error finding the correct tab" for now.

dvmarkusvogl reviewed (changes requested):
Looks fine functionally. Minor suggested clean code things.
Also: Any Tests / Documentation?

thinkh reviewed (changes requested):
Thanks for this nice refactoring! IMO it looks and feels already much better than the previous workbench transition.

I was focusing on the functional test for now. I'll do a code review afterwards. In the following are the points that I've found.

# Truncated workbench title of second workbench in 50:50 mode

**Steps to reproduce**
1. Open workbench (e.g., Drugs)
2. Select drugs
3. Open Cell Line workbench

**Observed behavior**

The title of the second workbench is truncated.

![grafik](https://user-images.githubusercontent.com/5851088/177939933-febf0431-ebcc-48b7-afa6-ca69ce0dd5ef.png)

**Expected behavior**

- [x] Show the full workbench title
- [ ] Left align title of second workbench to be consistent with the first workbench (?)
- [x] Show the `x of x <entity>` in 50:50 mode only if there is enough space (i.e, browser window is large enough); otherwise hide this information from both workbenches.


# Collapsed workbench chevron

**Steps to reproduce**
1. Open workbench (e.g., Drugs)
2. Select drugs
3. Open Cell Line workbench
4. Click on Drug workbench header to focus on that collapse the cell line workbench

**Observed behavior**

The chevron looks quite dense without margin. If you look closely then there is a fine white line between before the `>` triangle starts. 

![grafik](https://user-images.githubusercontent.com/5851088/177937428-49ac36a3-57e9-42da-a2e0-589872803c99.png)

**Expected behavior**

- [x] Make chevron a bit wider and give cell line a bit more padding to the left
- [ ] Add tool tip on hover to read the full workbench name (currently no tool tip available)
- [x] Change the chevron to look a bit more solid (without that fine white line)


# Wrong workbench layout on new analysis

**Steps to reproduce**

1. Open workbench (e.g., Drugs)
2. Open Cell Line workbench
3. Start a new analysis: Datasets -> Drugs

**Observed behavior**

The workbench layout after the restart is split in half.

![grafik](https://user-images.githubusercontent.com/5851088/177939060-454fa6f9-1e39-4142-9d55-1abd2a478912.png)

![ordino-wrong-workbench-layout](https://user-images.githubusercontent.com/5851088/177938691-9bc0ef06-70df-450e-858c-cb5255db6eef.gif)

This bug remains even when I start another analysis after the broken one. Only reloading the browser window helps. I assume that some flag in the redux store is not reset correctly.

**Expected behavior**

- [x] The workbench of a new analysis should start with full width.


# Third workbench

**Steps to reproduce**
1. Open workbench (e.g., Drugs)
2. Select drugs
3. Open Cell Line workbench
4. Select cell lines
5. Open Gene workbench

**Observed behavior**
 
The chevron of the first workbench is still visible. The side panels of the second workbench are open.

![grafik](https://user-images.githubusercontent.com/5851088/177944070-64ec1c5a-60a6-4fea-8896-ecc29cdd3c0b.png)


**Expected behavior**

- [x] Hide chevrons of the first workbench (i.e., focus - 1) in 50:50 mode
- [x] Add `Workbench` to context workbench title (see also my comment below)
- [ ] Close side panels of the second workbench, to provide more space for the rankings


# Further UI improvements

![grafik](https://user-images.githubusercontent.com/5851088/177940786-3989cf3c-36ee-444e-8e48-c3b2a83130f5.png)

- [ ] Align first workbench title with Ordino header and icons of sidebar
  

![grafik](https://user-images.githubusercontent.com/5851088/177941154-49fefc4f-038e-4dea-bc6e-39afc0f9a0db.png)

- [x] Left align context workbench title
- [x] Add `Workbench` to context workbench title
- [x] Align context splitter with chevron spacing


![grafik](https://user-images.githubusercontent.com/5851088/177941701-c7627174-a861-4e86-868a-974f2d18ea88.png)

- [ ] The `Add Views` looks a bit clumsy, I'd still like to see this text integrated in the chevron or at least style it a bit nicer


# Open questions / missing points

1. What has happened to the add views chevron we discussed with Alex last week?
   ![grafik](https://user-images.githubusercontent.com/5851088/177941638-8f91e405-9edc-4e8c-b242-2fab178bf524.png)
   ![grafik](https://user-images.githubusercontent.com/5851088/177942222-42bf20aa-903a-4d4a-82b7-3aa959caee87.png)
   ![grafik](https://user-images.githubusercontent.com/5851088/177942432-beb0be88-406a-4f17-8c29-374aefc81fe3.png)
2. What has happened to the colored indicators in the detail view chooser?
   ![grafik](https://user-images.githubusercontent.com/5851088/177942854-8696a82a-af92-4990-bac6-9e628971c629.png)

dvzacharycutler commented:
Holger and I had a discussion about the open questions. For now, the first question (add views chevron) we will leave out and try out other ways to make the view chooser more obvious in a future PR. And the colored indicator will also be a part of a future PR. otherwise, I addressed all of the issues brought up by holger.

dvzacharycutler commented:
> Also: Any Tests / Documentation?

Good question, the tests that would ideally come with this at the moment is storybook examples for the new components that are being created with this PR. But setting up storybook within Ordino is a bit harder than anticipated, so although I made storybook files for the new components (WorkbenchUtilsSidebar.stories.tsx, for example), they error when running storybook. I believe Holger is taking a look at how to fix it.

thinkh commented on src/components/breadcrumb/Breadcrumb.tsx:
```diff
@@ -0,0 +1,85 @@
+    return counter;
+  }, [ordino.workbenches.length, ordino.focusWorkbenchIndex]);
+
+  console.log(endFlexNum);
```
```suggestion
```

dvzacharycutler replied:
done

thinkh commented on package.json:
```diff
@@ -88,11 +90,13 @@
     "@blueprintjs/core": "^4.2.1",
     "@blueprintjs/icons": "^4.2.4",
     "@lineup-lite/table": "1.6.0",
+    "@storybook/react": "^6.4.19",
```
I think this dependency should be a `devDependency` because we do not use the types for the production build.

dvzacharycutler replied:
done

thinkh reviewed (commented):
Thanks for addressing already some feedback from above. I will do another functional test later. Here are some comments from my code review.

dvzacharycutler commented on src/components/breadcrumb/BreadcrumbSvg.tsx:
```diff
@@ -0,0 +1,80 @@
+  first = false,
+  margin = 4,
+  color = 'cornflowerblue',
+  clickable = false,
```
done

dvzacharycutler commented on src/components/breadcrumb/BreadcrumbSvg.tsx:
```diff
@@ -0,0 +1,80 @@
+  width = 1500,
+  height = 40,
+  chevronIndent = 8,
+  first = false,
```
Done

dvzacharycutler commented on src/app/viewChooser/ViewChooser.tsx:
```diff
@@ -139,6 +139,7 @@ export function ViewChooser({
         <div className={`view-chooser-content bg-white d-flex flex-column justify-content-stretch ${!embedded && !collapsed ? 'shadow' : ''}`}>
           {showHeader && (
             <ViewChooserHeader>
+              <div className="text-gray h4">Add Views</div>
```
Deleted this text anyways

dvzacharycutler commented on src/components/breadcrumb/Breadcrumb.tsx:
```diff
@@ -0,0 +1,84 @@
+
+export const colorPalette = ['#337ab7', '#ec6836', '#75c4c2', '#e9d36c', '#24b466', '#e891ae', '#db933c', '#b08aa6', '#8a6044', '#7b7b7b'];
+
+const SMALL_CHEVRON_WIDTH = 3;
```
Done

dvzacharycutler commented on src/components/breadcrumb/BreadcrumbSvg.tsx:
```diff
@@ -0,0 +1,80 @@
+  margin = 4,
+  color = 'cornflowerblue',
+  clickable = false,
+}: {
+  width?: number;
+  chevronIndent?: number;
+  first?: boolean;
+  height?: number;
+  color?: string;
+  margin?: number;
+  clickable?: boolean;
+}) {
```
I've actually been wondering this, Michael has given me the impression in the past he prefers doing these typings in line instead of in an interface. 

So I kept it inline, but added some comments

dvzacharycutler commented on src/components/breadcrumb/BreadcrumbSvg.tsx:
```diff
@@ -0,0 +1,80 @@
+  return `#${(g | (b << 8) | (r << 16)).toString(16)}`;
+}
+export function BreadcrumbSvg({
+  width = 1500,
+  height = 40,
+  chevronIndent = 8,
```
Done

dvzacharycutler commented on src/components/breadcrumb/BreadcrumbSvg.tsx:
```diff
@@ -0,0 +1,80 @@
+  height = 40,
+  chevronIndent = 8,
+  first = false,
+  margin = 4,
```
Done

dvzacharycutler commented on src/components/breadcrumb/Breadcrumb.tsx:
```diff
@@ -0,0 +1,84 @@
+              flexWidth = ordino.midTransition ? CHEVRON_TRANSITION_WIDTH : SMALL_CHEVRON_WIDTH;
+            } else {
+              flexWidth = SMALL_CHEVRON_WIDTH;
+            }
```
Done

thinkh commented on src/scss/components/ChevronBreadcrumb.scss:
```diff
@@ -23,9 +23,18 @@
 }
 
-.chevronDiv span, p {
+.ellipsisText {
```
Could you use [`text-truncate` from Bootstrap](https://getbootstrap.com/docs/5.1/helpers/text-truncation/)?

thinkh commented on src/app/viewChooser/ViewChooserFilter.tsx:
```diff
@@ -20,7 +20,7 @@ export function ViewChooserFilter(props: IViewChooserFilterProps) {
   return (
     <div className="view-filter input-group flex-nowrap">
-      <input className="form-control border-end-0" type="search" placeholder="Search" value={filter} onChange={(evt) => setFilter(evt.target.value)} />
+      <input className="form-control border-end-0" type="search" placeholder="Search Views" value={filter} onChange={(evt) => setFilter(evt.target.value)} />
```
```suggestion
      <input className="form-control border-end-0" type="search" placeholder="Search views" value={filter} onChange={(evt) => setFilter(evt.target.value)} />
```

Use sentence case according to style guide.

thinkh commented on src/scss/components/Workbench.scss:
```diff
@@ -60,4 +60,8 @@
   border-right: 1px solid #dee2e6;
+}
+
+.borderRadiusNone {
```
Could you use [`rounded-0` from Bootstrap](https://getbootstrap.com/docs/5.1/utilities/borders/#sizes)?

dvzacharycutler replied:
yep! changed and deleted borderRadiusNone class

thinkh reviewed (commented):
Thanks for your changes and comments in the source code. Please have a look if you can use Bootstrap classes and use sentence case. I will do a functional test tomorrow.

thinkh reviewed (changes requested):
In the following the results of my functional test.

The text in the login screen is currently broken due to a `nowrap` style.

![image](https://user-images.githubusercontent.com/5851088/178427767-7a986e76-b9d8-42d9-b401-8676f8aaf7bd.png)

I suggest to use the workbench color as active color in the sidebar.

![image](https://user-images.githubusercontent.com/5851088/178431951-5f108d0f-8850-4d9f-a594-0fcf995d31ef.png)

Will continue later and keep you posted.

dvzacharycutler commented:
> The text in the login screen is currently broken due to a `nowrap` style.

Fixed.

> I suggest to use the workbench color as active color in the sidebar.

Done, changed the colors in the checkbox as well. 

Also fixed the problems mentioned in the most recent comment by holger, except for the last one. We discussed, and ill make some design changes of shrinking the non-immediate chevrons down to be very small and have no text, but in the next PR.

thinkh commented:
@dvzacharycutler Thanks for your last iteration. It looks good very good!

I fixed the Storybook SCSS issue. It requires the export in tdp_core with PR https://github.com/datavisyn/tdp_core/pull/753. The Storybook server does start now and you can open the Storybook UI in the browser. However, the story/stories that are already available do not compile due to some React Redux error.

![image](https://user-images.githubusercontent.com/5851088/178679043-7185bb3e-0400-40c7-a7ce-bf24425cebb7.png)

```
could not find react-redux context value; please ensure the component is wrapped in a <Provider>
Error: could not find react-redux context value; please ensure the component is wrapped in a <Provider>
    at useReduxContext (http://localhost:6006/vendors~main.iframe.bundle.js:214802:11)
    at useSelector (http://localhost:6006/vendors~main.iframe.bundle.js:214951:28)
    at WorkbenchUtilsSidebar (http://localhost:6006/main.iframe.bundle.js:38662:101)
    at renderWithHooks (http://localhost:6006/vendors~main.iframe.bundle.js:199851:18)
    at mountIndeterminateComponent (http://localhost:6006/vendors~main.iframe.bundle.js:202530:13)
    at beginWork (http://localhost:6006/vendors~main.iframe.bundle.js:203644:16)
    at HTMLUnknownElement.callCallback (http://localhost:6006/vendors~main.iframe.bundle.js:185236:14)
    at Object.invokeGuardedCallbackDev (http://localhost:6006/vendors~main.iframe.bundle.js:185285:16)
    at invokeGuardedCallback (http://localhost:6006/vendors~main.iframe.bundle.js:185340:31)
    at beginWork$1 (http://localhost:6006/vendors~main.iframe.bundle.js:208251:7)
```

Can you have a look, if this is easily fixable? Otherwise, I'd suggest to take a closer look as part of a separate issue, as this might trigger refactorings of our components.

thinkh reviewed (approved):
I agreed with @dvzacharycutler that he will fix the storybook stories with the next PR #412. Since everything has been addressed, I'll merge this PR.
